query-planner/src/model.rs

```diff
@@ -0,0 +1,67 @@
+//! This is the model object for a QueryPlan trimmed down to only contain _owned_ fields
+//! that are required for executing a QueryPlan as implemented in the existing Apollo Gateway.
+//!
+//! The [SelectionSet] in the `requires` field of a [FetchNode] is trimmed to only be a list of
+//! either a [Field] or an [InlineFragment], since those are the only potential values needed to
+//! execute a query plan. Furthermore, within a [Field] or [InlineFragment], we only need
+//! names, aliases, type conditions, and recurively sub [SelectionSet]s.
+
+use std::fmt;
+
+#[derive(Debug, PartialEq)]
+pub struct QueryPlan(pub Option<PlanNode>);
+
+#[derive(Debug, PartialEq)]
+pub enum PlanNode {
+    Sequence(Vec<PlanNode>),
+    Parallel(Vec<PlanNode>),
+    Fetch(FetchNode),
+    Flatten {
+        path: Vec<ResponsePathElement>,
+        node: Box<PlanNode>,
+    },
+}
+
+#[derive(Debug, PartialEq)]
+pub struct FetchNode {
+    pub service_name: String,
+    pub variable_usages: Vec<String>,
+    pub requires: Option<SelectionSet>,
+    pub operation: String,
```

Comment:
To be clear, this is the minified printed GraphQL Document right? Is there a way to model that in the typesystem here?

Reply:
Well, you tell me? In JS this is how we create the value for this field (`source`)
```
const operation = requires
  ? operationForEntitiesFetch({
      selectionSet,
      variableUsages,
      internalFragments,
    })
  : operationForRootFetch({
      selectionSet,
      variableUsages,
      internalFragments,
      operation: context.operation.operation,
    });

...
source: stripIgnoredCharacters(print(operation))
```

If I'm reading it correctly, this field encapsulates the operation we're going to send for this fetch node; it's built by processing the selectionSet, fragments and variables for this fetch node. Those fields have been removed from the Rust `QueryPlan` model because they're not used for query execution. Assuming we'll want to cache these instances of `QueryPlan`, then if we want to capture these values in the type system, we'll have to cache many values from the AST nodes, anything that ends up being part of the operation string. Then when we load a query from cache, we'll have to recalculate the fetch operation.